File: RedditPictureReader/app/src/main/java/se/slotte/ludwig/redditpicturereader/favourite_pictures/adapter/MyFavouritePictureRecyclerViewAdapter.java
package se.slotte.ludwig.redditpicturereader.favourite_pictures.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.like.LikeButton;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.Realm;
import se.slotte.ludwig.redditpicturereader.R;
import se.slotte.ludwig.redditpicturereader.shared.realm.model.RealmSavedFavouritePicture;

/**
 * Created by ludwigslotte on 15/10/16.
 */

public class MyFavouritePictureRecyclerViewAdapter extends RecyclerView.Adapter<MyFavouritePictureRecyclerViewAdapter.ViewHolder> {
    private List<RealmSavedFavouritePicture> photoList;
    private Context context;

    public MyFavouritePictureRecyclerViewAdapter() {
        photoList = new ArrayList<>();
    }

    public void setItems(List<RealmSavedFavouritePicture> list) {
        photoList = list;
        notifyDataSetChanged();
    }

    @Override
    public MyFavouritePictureRecyclerViewAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.favourite_picture_item, parent, false);
        context = parent.getContext();
        return new MyFavouritePictureRecyclerViewAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final MyFavouritePictureRecyclerViewAdapter.ViewHolder holder, final int position) {
        final Realm realm = Realm.getDefaultInstance();
        final RealmSavedFavouritePicture picture = photoList.get(position);

        holder.thumbnail.setVisibility(View.VISIBLE);
        holder.starButton.setVisibility(View.VISIBLE);
        holder.starButton.setLiked(true);

        FavouritePictureAdapterHelper helper = new FavouritePictureAdapterHelper(holder, picture);
        helper.loadGlide(context);
        helper.starButtonLogic(this, photoList, position, realm);
    }

    @Override
    public int getItemCount() {
        return photoList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        final View mView;
        @BindView(R.id.thumbnail)
        ImageView thumbnail;
        @BindView(R.id.star_button)
        LikeButton starButton;

        ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
            mView = view;
        }

    }
}

File: RedditPictureReader/app/src/main/java/se/slotte/ludwig/redditpicturereader/favourite_pictures/adapter/MyFavouritePictureRecyclerViewAdapter.kt
package se.slotte.ludwig.redditpicturereader.favourite_pictures.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.like.LikeButton
import io.realm.Realm
import kotlinx.android.synthetic.main.favourite_picture_item.view.*
import se.slotte.ludwig.redditpicturereader.R
import se.slotte.ludwig.redditpicturereader.didChange
import se.slotte.ludwig.redditpicturereader.favourite_pictures.adapter.MyFavouritePictureRecyclerViewAdapter.ViewHolder
import se.slotte.ludwig.redditpicturereader.shared.realm.model.RealmSavedFavouritePicture
import kotlin.properties.Delegates


class MyFavouritePictureRecyclerViewAdapter : RecyclerView.Adapter<ViewHolder>() {
    private lateinit var context: Context
    var photoList by Delegates.didChange(emptyList<RealmSavedFavouritePicture>()) { notifyDataSetChanged() }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.favourite_picture_item, parent, false)
        context = parent.context
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val realm = Realm.getDefaultInstance()
        val picture = photoList[position]

        holder.apply {
            thumbnail?.visibility = View.VISIBLE
            starButton?.visibility = View.VISIBLE
            starButton?.setLiked(true)
        }

        FavouritePictureAdapterHelper(holder, picture).apply {
            loadGlide(context)
            starButtonLogic(this@MyFavouritePictureRecyclerViewAdapter, photoList, position, realm)
        }
    }

    override fun getItemCount(): Int {
        return photoList.size
    }

    inner class ViewHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {
        internal var thumbnail: ImageView? = view.thumbnail
        internal var starButton: LikeButton? = view.star_button
    }
}